privval/test_harness.go

```diff
@@ -0,0 +1,263 @@
+	BindAddr string
+
+	KeyFile     string
+	StateFile   string
+	GenesisFile string
+
+	AcceptDeadline time.Duration
+	ConnDeadline   time.Duration
+
+	SecretConnKey ed25519.PrivKeyEd25519
+}
+
+// NewTestHarness will load Tendermint data from the given files (including
+// validator public/private keypairs and chain details) and create a new
+// harness.
+func NewTestHarness(logger log.Logger, cfg TestHarnessConfig) (*TestHarness, error) {
+	var err error
+	var keyFile, stateFile, genesisFile string
+
+	if keyFile, err = expandPath(cfg.KeyFile); err != nil {
+		return nil, newTestHarnessError(ErrFailedToExpandPath, err, cfg.KeyFile)
+	}
+	if stateFile, err = expandPath(cfg.StateFile); err != nil {
+		return nil, newTestHarnessError(ErrFailedToExpandPath, err, cfg.StateFile)
+	}
+	logger.Info("Loading private validator configuration", "keyFile", keyFile, "stateFile", stateFile)
+	// NOTE: LoadFilePV ultimately calls os.Exit on failure. No error will be
+	// returned if this call fails.
+	fpv := LoadFilePV(keyFile, stateFile)
+
+	if genesisFile, err = expandPath(cfg.GenesisFile); err != nil {
+		return nil, newTestHarnessError(ErrFailedToExpandPath, err, cfg.GenesisFile)
+	}
+	logger.Info("Loading chain ID from genesis file", "genesisFile", genesisFile)
+	st, err := state.MakeGenesisDocFromFile(genesisFile)
+	if err != nil {
+		return nil, newTestHarnessError(ErrFailedToLoadGenesisFile, err, genesisFile)
+	}
+	logger.Info("Loaded genesis file", "chainID", st.ChainID)
+
+	sc, err := newTestHarnessSocketVal(logger, cfg)
+	if err != nil {
+		return nil, newTestHarnessError(ErrFailedToCreateListener, err, "")
+	}
+
+	return &TestHarness{
+		sc:      sc,
+		fpv:     fpv,
+		chainID: st.ChainID,
+		logger:  logger,
+	}, nil
+}
+
+func (th *TestHarness) Run() {
+	th.logger.Info("Starting test harness")
+	donec := make(chan struct{})
+	go func() {
+		defer close(donec)
+		if err := th.sc.Start(); err != nil {
+			th.logger.Error("Failed to start listener", "err", err)
+			th.Shutdown(newTestHarnessError(ErrFailedToStartListener, err, ""))
+		}
+		th.TestPublicKey()
+		th.TestSignProposal()
+	}()
+
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt)
+	go func() {
+		for sig := range c {
+			th.logger.Info("Caught interrupt, terminating...", "sig", sig)
+			th.Shutdown(newTestHarnessError(ErrInterrupted, nil, ""))
+		}
+	}()
+
+	// Run until complete
+	<-donec
+	th.Shutdown(nil)
+}
+
+func (th *TestHarness) TestPublicKey() {
+	th.logger.Info("TEST: Public key of remote signer")
+	th.logger.Info("Local", "pubKey", th.fpv.GetPubKey())
+	th.logger.Info("Remote", "pubKey", th.sc.GetPubKey())
+	if th.fpv.GetPubKey() != th.sc.GetPubKey() {
+		th.logger.Error("FAILED: Local and remote public keys do not match")
+		th.Shutdown(newTestHarnessError(ErrTestPublicKeyFailed, nil, ""))
+	}
+}
+
+func (th *TestHarness) TestSignProposal() {
+	th.logger.Info("TEST: Signing of proposals")
+	prop := &types.Proposal{Timestamp: time.Now()}
```

Comment:
This won't pass validation in the kms as this does'nt represent a valid proposal:
https://github.com/tendermint/kms/blob/b4cb33d7255dfb6f4245af8ce39a1feb41299023/tendermint-rs/src/amino_types/proposal.rs#L139-L159

Reply:
Oh I see. I haven't gotten to grips yet with the internals of the proposal mechanism. Let me take a look at that, thanks! 👍

Reply:
Or on the tendermint side:
https://github.com/tendermint/tendermint/blob/aa40cfcbb97fac3832b47ccdcf99c6b7fb504cb2/types/proposal.go#L47-L50

Reply:
Adding a `Type` field, along with a `Height` and `Round` value, now successfully signs the proposal 👍 I now just have to validate the signature from the test harness' side, and then we'll have the first cryptographic test of the KMS by the harness 👶